### ISSUE TITLE
connecting to raspberry pi

### ISSUE THREAD

**Issue by slowphil** [TRMC2/libtrmc2#1]
First of all a big thank you for sharing this! We still have a few TRMC2 in use in the lab, and it's a headache whenever we need to replace the old PC they work with. Using a Raspberry Pi for that (perhaps inside the TRMC2 box itself) is something I have been dreaming of for a long time.

Now to the point: In practice, how do you connect the Trmc2 to a Raspberry?

AFAIK, when using with a PC we need some sort of cumbersome adapter connected to both a DB9 serial port a mini-din-like serial mouse port... Is this adapter (or another one) still needed with the Raspberry? And which GPIO pins do you use (although I guess I could find it in the code)? 

In short, if you had a schematics for connecting to the Pi, that would be super.

**Comment by edgar-bonet** [comment]
I am glad this can be useful! Just out of curiosity, which lab is yours?

I have not made this connection myself, and I am not quite sure about what is needed. At the very least, you will need a level-converter chip suitable for RS-232 voltage levels. There may be something more needed depending on the version of the TRMC2. For example, not all TRMC2s need the mini-DIN link. I will hunt down the people who know and get back to you...

Anyway, you are making a very good point: this should be documented.

As for the pin numbers, it is indeed [in the code][1]:

```c
# define PIN_CLOCK      17  /* GPIO17, pin 11 */
# define PIN_DATA_IN    18  /* GPIO18, pin 12 */
# define PIN_DATA_OUT   27  /* GPIO27, pin 13 */
```

[1]: ../blob/master/TrmcLin.c#L47-L49

**Comment by edgar-bonet** [comment]
> Is this adapter (or another one) still needed with the Raspberry?

Some sort of adapter will be needed for sure, as the GPIO pins of the Raspberry Pi will not be able to drive the TRMC2 directly. The straightforward solution is to build an adapter that would be used _in addition_ to your current adapter. The idea would be to make a drop-in replacement for your current PC. For this, you new adapter would need:

 * a male DE-9 connector to replace the serial port of the PC

 * a female PS/2 connector to replace the mouse port of the PC.

The PS/2 connector is only used to provide a low-current +5V supply. It would be connected to GND and +5V pins on the Raspberry Pi. The DE-9 connector is used for the data link. Only pins 4 (DTR), 7 (RTS), 8 (CTS) and 5 (GND) are used, but the signal levels must be RS-232-compatible. You will thus need a transceiver, such as the MAX3232, for converting between the RS-232 levels and the 3.3V logic of the Raspberry Pi.

Below is a simplified schematic of the adapter. Note that I omitted the capacitors that are typically needed to make an RS-232 transceiver work:

                                       Raspberry Pi
                        +3.3V             ┏━━━━┓
      DE-9                ┌───────────────╂─01 ┃
    connector      ┏━━━━━━┷━━━━━━┓        ┃    ┃
      ┏━━━┓        ┃      ╱│     ┃        ┃    ┃
      ┃ 4─╂────────╂────o⟨ ├─────╂────────╂─11 ┃
      ┃   ┃        ┃      ╲│     ┃        ┃    ┃
      ┃   ┃        ┃             ┃        ┃    ┃
      ┃   ┃        ┃      ╱│     ┃        ┃    ┃
      ┃ 7─╂────────╂────o⟨ ├─────╂────────╂─13 ┃
      ┃   ┃        ┃      ╲│     ┃        ┃    ┃
      ┃   ┃        ┃             ┃        ┃    ┃
      ┃   ┃        ┃     │╲      ┃        ┃    ┃
      ┃ 8─╂────────╂─────┤ ⟩o────╂────────╂─12 ┃
      ┃   ┃        ┃     │╱      ┃        ┃    ┃
      ┃   ┃        ┗━━━━━━┯━━━━━━┛        ┃    ┃
      ┃ 5─╂─────┐         │               ┃    ┃
      ┗━━━┛     ├─────────┴───────────────╂─09 ┃
      ┏━━━┓     │        GND              ┃    ┃
      ┃ 3─╂─────┘                         ┃    ┃
      ┃   ┃              +5V              ┃    ┃
      ┃ 4─╂───────────────────────────────╂─02 ┃
      ┗━━━┛                               ┗━━━━┛
      PS/2
    connector

You have some leeway for the pin assignments on the Raspberry Pi:

 * You can use either pin 2 or 4 to provide +5V.

 * You can use any ground pin for the ground.

 * You could also change the data pins, if you adjust the `#define` lines in the code accordingly. Just make sure to select pins that are not otherwise used: see the output of `gpioinfo`.

Although this should work, it is not the most elegant solution: you would have two adapters chaining voltage level conversions. It should be possible to combine them into one adapter and have a single stage of level shifting. We are digging out the old schematics of the TRMC2 in order to find out what the requirements of such level conversion would be.

**Comment by slowphil** [comment]
Thanks a lot for the very detailed reply. Right now I have it running on a linux PC, but will switch asap to a RPI.
Long live TRMC2 in Saclay!
